### PR TITLE
support install into java cacerts file

### PR DESCRIPTION
**Pull request by adamdecaf**
This works for me locally on a Mac with Java 8 and 9. I'm going to test out a bit on linux and newer Java versions. 

```
$ ./bin/mkcert -install
Using the local CA at "/Users/adam/Library/Application Support/mkcert" ✨
The local CA is now installed in the system trust store! ⚡️
The local CA is now installed in the Firefox trust store (requires browser restart)! 🦊
The local CA is now installed in Java's trust store ☕️

$ keytool -list -keystore $JAVA_HOME/jre/lib/security/cacerts -storepass changeit | grep -i mkcert
mkcert development ca 215883752978067404179277978616948247614, Jul 6, 2018, trustedCertEntry, 
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -194,6 +203,15 @@ func (m *mkcert) uninstall() {
 			log.Print("")
 		}
 	}
+	if hasJava {
+		if hasKeytool {
+			m.uninstallJava()
+		} else {
+			log.Print("")
+			log.Printf(`Warning: "keytool" is not available, so the CA can't be automatically uninstalled from %s (if it was ever installed)! ⚠️`)
```
Comment (FiloSottile) [inline main.go]: Leftover `%s`

File: main.go [hunk 2]
```diff
@@ -178,6 +182,11 @@ func (m *mkcert) install() {
 		}
 		printed = true
 	}
+	if hasJava && !m.checkJava() {
+		m.installJava()
+		log.Println("The local CA is now installed in Java's trust store 🦊")
```
Comment (FiloSottile) [inline main.go]: ☕️

File: main.go [hunk 3]
```diff
@@ -178,6 +182,11 @@ func (m *mkcert) install() {
 		}
 		printed = true
 	}
+	if hasJava && !m.checkJava() {
+		m.installJava()
```
Comment (FiloSottile) [inline main.go]: Shouldn't this be behind a `if hasKeytool` like `if hasCertutil` above?

File: truststore_java.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	hasKeytool bool
+
+	javaHome    string
+	cacertsPath string
+	keytoolPath string
+	storePass   string = "changeit"
+)
+
+func init() {
+	if v := os.Getenv("JAVA_HOME"); v != "" {
```
Comment (FiloSottile) [inline truststore_java.go]: How common/expected is it for `JAVA_HOME` to be set in the default environment?

File: truststore_java.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	// is in the keytool -list output
+	exists := func(c *x509.Certificate, h hash.Hash, keytoolOutput []byte) bool {
+		_, err := h.Write(c.Raw)
+		if err != nil {
+			return false
+		}
+		// keytool outputs SHA1 and SHA256 certificates in uppercase hex
+		// with each octet pair deliniated by ":"
+		fp := []byte(strings.ToUpper(hex.EncodeToString(h.Sum(nil))))
+		return bytes.Contains(keytoolOutput, fp)
```
Comment (FiloSottile) [inline truststore_java.go]: Nit: I would move the `[]byte` conversion to this line, as `fp` makes sense as a string, and we are converting it for the benefit of `bytes.Contains`.

File: truststore_java.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	keytoolOutput, err := exec.Command(keytoolPath, "-list", "-keystore", cacertsPath, "-storepass", storePass).CombinedOutput()
+	if err != nil {
+		hasKeytool = false
```
Comment (FiloSottile) [inline truststore_java.go]: Why? Could we catch it in `init()`?

File: truststore_java.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+func (m *mkcert) checkJava() bool {
+	// exists returns true if the given x509.Certificate's fingerprint
+	// is in the keytool -list output
+	exists := func(c *x509.Certificate, h hash.Hash, keytoolOutput []byte) bool {
+		_, err := h.Write(c.Raw)
+		if err != nil {
+			return false
+		}
+		// keytool outputs SHA1 and SHA256 certificates in uppercase hex
+		// with each octet pair deliniated by ":"
```
Comment (FiloSottile) [inline truststore_java.go]: "delimitated"

But also, the ":" are gone by the time we get here, so this comment is confusing.

File: truststore_java.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+// execKeytool will execute a "keytool" command and if needed re-execute
+// the command wrapped in 'sudo' to work around file permissions.
+func (m *mkcert) execKeytool(cmd *exec.Cmd) ([]byte, error) {
+	out, err := cmd.CombinedOutput()
+	if err != nil && bytes.Contains(out, []byte("java.io.FileNotFoundException")) {
+		origArgs := cmd.Args
+		cmd = exec.Command("sudo", keytoolPath)
+		cmd.Args = append(cmd.Args, origArgs...)
+		cmd.Env = []string{
+			fmt.Sprintf("JAVA_HOME=%s", javaHome),
```
Comment (FiloSottile) [inline truststore_java.go]: Nit: `javaHome` is a string, no need for `fmt.Sprintf`, just use `+`.

File: truststore_java.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	out, err := m.execKeytool(exec.Command(keytoolPath, args...))
+	fatalIfCmdErr(err, "keytool -importcert", out)
+}
+
+func (m *mkcert) uninstallJava() {
+	args := []string{
+		"-delete",
+		"-alias", fmt.Sprintf("'%s'", m.caUniqueName()),
```
Comment (FiloSottile) [inline truststore_java.go]: Why the quotes?

File: truststore_java.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+func (m *mkcert) installJava() {
+	tmp, err := ioutil.TempFile("", "mkcert-java")
+	fatalIfErr(err, "failed creating temp file for java install")
+	defer os.Remove(tmp.Name())
+
+	err = ioutil.WriteFile(tmp.Name(), pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: m.caCert.Raw,
+	}), 0644)
+	fatalIfErr(err, "failed to write java cert into temp file")
```
Comment (FiloSottile) [inline truststore_java.go]: Could you use the certificate file on disk? See for example the macOS `security` input.

File: truststore_java.go (hunk 4 above)
Comment (adamdecaf) [reply truststore_java.go]: On servers (as the user running java apps) it should exist. Developers _should_ have JAVA_HOME set, but they might not.

It's probably good enough for the 80% usecase.

**Comment by adamdecaf** [comment]
@FiloSottile Updated! ☕️

**Comment by FiloSottile** [comment]
BTW, GitHub code reviews work best if you upload revisions as commits, the opposite of Gerrit. I squash on merge anyway.

**Comment by adamdecaf** [comment]
@FiloSottile Yea. I don't usually like those intermediate commits. 

Do you see anything else for this? 

I've used `/etc/alternatives/java` (on ubuntu) to find java before. (Symlink to install dir)

File: truststore_java.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+	out, err := m.execKeytool(exec.Command(keytoolPath, args...))
+	if bytes.Contains(out, []byte("does not exist")) {
+		return // cert didn't exist
+	}
+	fatalIfCmdErr(err, "keytool -delete", out)
+}
+
+// execKeytool will execute a "keytool" command and if needed re-execute
+// the command wrapped in 'sudo' to work around file permissions.
+func (m *mkcert) execKeytool(cmd *exec.Cmd) ([]byte, error) {
```
Comment (adamdecaf) [inline truststore_java.go]: I see https://github.com/FiloSottile/mkcert/commit/564b41305c34b37318c3f3130c32d87980af074e added CommandWithSudo. I'll switch to that here.

Comment (adamdecaf) [reply truststore_java.go]: Eh, maybe not. CommandWithSudo isn't quite a drop in replacement.

**Comment by adamdecaf** [comment]
@FiloSottile Did you notice anything else todo on this?

**Pull request merged**